### PR TITLE
fix: harden regex builtins with compiled-size limit

### PR DESCRIPTION
**Pull request by anakrish**
## Summary

Adds a compiled-size limit to all regex builtins in Regorus, preventing adversarial patterns from consuming excessive CPU during DFA construction or matching.

## Approach

- **Engine-level cap**: Uses `RegexBuilder::size_limit(100KB)` to reject patterns whose compiled NFA exceeds the limit
- **Centralized helper**: `compile_regex_for_builtin()` applies the limit consistently across all regex builtins
- **Error propagation**: `LimitError::RegexSizeLimitExceeded` propagates as a fatal VM error that cannot be absorbed by rule evaluation (including under negation)
- **State cleanup**: `restore_rule_state()` ensures register/loop/comprehension stacks are properly unwound on fatal errors

## Why 100KB?

- All real-world policy patterns observed in production compile well under 100KB
- All adversarial patterns identified during benchmarking exceed this threshold
- Making the limit configurable at runtime can be added as a follow-up if needed

## Pre-existing bug fix

Resource-limit errors (time, memory, instruction count) raised inside builtins were silently swallowed to `Undefined` when `strict_builtin_errors` was off (the default). This meant `not regex.match(...)` with an oversized pattern would see `Undefined` and flip to `true` — silently wrong. Fixed by teaching the error-absorption paths in the interpreter, RVM builtin dispatch, and RVM rule-execution loop to recognize `LimitError` and propagate it.

## Test coverage

Regression tests cover legitimate policy patterns (IPv4, hostname, semver, UUID, email, ARM resource IDs, etc.) alongside adversarial patterns that must be rejected. Edge cases include negation (`not regex.match`), and oversized patterns through every regex builtin: `regex.replace`, `regex.find_n`, `regex.find_all_string_submatch_n`, `regex.split`, `regex.template_match`, and `regex.is_valid`. The test harness runs each case through both the tree-walking interpreter and the RVM.

## Related

- Adversarial benchmark gist: https://gist.github.com/anakrish/832831bec3469b0e8c729ff8b825a049
- Follow-up issue for json.match_schema bypass: #706

### CODE REVIEW

File: src/rvm/vm/rules.rs [hunk 1]
```diff
@@ -79,6 +91,7 @@ impl RegoVM {
                 {
                     match self.jump_to(destructuring_entry_point) {
                         Ok(_result) => {}
+                        Err(e) if Self::is_fatal_vm_error(&e) => return Err(e),
                         Err(_e) => {
```
[the remark this answers is not in the document]
Comment (anakrish) [reply src/rvm/vm/rules.rs:119]: Fixed — added a `restore_rule_state()` helper that restores `register_stack` (pop + swap), `loop_stack`, and `comprehension_stack` before the early return. Both the destructuring path (line 113) and the body path (line 151) now call it.

The `call_rule_stack` entry is popped by the caller (`execute_call_rule_common`) on error propagation, so that was already correct.

File: src/rvm/vm/rules.rs [hunk 2]
```diff
@@ -111,6 +124,7 @@ impl RegoVM {
                         // are treated as else-branches and must not be evaluated.
                         break;
                     }
+                    Err(e) if Self::is_fatal_vm_error(&e) => return Err(e),
                     Err(_e) => {}
```
[the remark this answers is not in the document]
Comment (anakrish) [reply src/rvm/vm/rules.rs:158]: Same fix as above — `restore_rule_state()` is called before this early return as well (line 152-155). Both paths now properly unwind the swapped state before propagating the fatal error.

File: src/rvm/vm/execution.rs [hunk 3]
```diff
@@ -570,7 +570,13 @@ impl RegoVM {
         Ok(())
     }
 
-    fn handle_instruction_error(&mut self, _err: VmError, last_result: &mut Value) -> Result<bool> {
+    fn handle_instruction_error(&mut self, err: VmError, last_result: &mut Value) -> Result<bool> {
+        // Resource-limit errors must never be absorbed by rule evaluation.
+        // They represent engine-level constraints, not rule-level failures.
+        if RegoVM::is_fatal_vm_error(&err) {
+            return Ok(false);
```
[the remark this answers is not in the document]
Comment (anakrish) [reply src/rvm/vm/execution.rs:579]: This is intentional and safe. When `handle_instruction_error` returns `Ok(false)` for a fatal error, the caller at line 444-446 does:
```rust
self.execution_stack.clear();
return Err(err);
```

The error propagates all the way up through `run_stackless_loop` → `execute_rule`/`resume_execution`, terminating the evaluation. After a fatal resource-limit error, the VM instance is not reused — the error surfaces to the caller who discards the evaluation context entirely.

Added a clarifying comment documenting this contract.

File: src/rvm/vm/errors.rs [hunk 4]
```diff
@@ -298,6 +301,29 @@ pub enum VmError {
...
+                    VmError::MemoryLimitExceeded {
+                        usage,
+                        limit,
+                        pc: 0,
+                    }
+                }
+                crate::LimitError::RegexSizeLimitExceeded { limit } => {
+                    VmError::RegexSizeLimitExceeded { limit, pc: 0 }
+                }
+            };
```
[the remark this answers is not in the document]
Comment (anakrish) [reply src/rvm/vm/errors.rs:328]: This is a structural limitation of `From<anyhow::Error>` — there is no instruction context available at the conversion site. The pre-existing `ArithmeticError` path (line 330) has the same `pc: 0` for the same reason.

In practice this is acceptable because:
1. The error message itself (e.g. "regex compiled size exceeds limit of 100KB") tells the user exactly what went wrong
2. The pattern string in the policy source is the actionable fix target, not the RVM instruction offset
3. Threading `pc` through `anyhow::Error` would require invasive changes to the builtin calling convention for minimal diagnostic gain

Added a comment documenting why `pc: 0` is expected here.

File: tests/rvm/rego/cases/regex_size_limit.yaml [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+    query: data.test.main
+    want_result: false
+
+  # -------------------------------------------------------------------------
+  # Adversarial patterns — must be rejected as hard errors.
+  #
+  # These patterns produce compiled NFAs exceeding the 100 KiB size limit.
+  # The error must propagate (not be swallowed to undefined) so that
+  # `not regex.match(...)` does not silently flip to true.
+  # -------------------------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (anakrish) [reply tests/rvm/rego/cases/regex_size_limit.yaml:160]: Added two new test cases:
- `blocked-not-regex-match-oversized`: verifies that `not regex.match(oversized, input)` produces an error rather than silently evaluating to `true`
- `blocked-regex-replace-oversized`: covers the `regex.replace` path with an oversized pattern

Regarding suspendable-mode coverage: the RVM test harness runs in default (run-to-completion) mode. The error propagation through `handle_instruction_error` → `execution_stack.clear()` → `Err(e)` is the same codepath regardless of whether suspension was previously active — the fatal error terminates evaluation unconditionally. A dedicated suspendable-mode test would require a multi-step evaluation scenario which is beyond the scope of this hardening PR.

File: src/builtins/regex.rs [hunk 6]
```diff
@@ -185,6 +224,13 @@ fn regex_replace(
...
         Ok(p) => p,
+        Err(regex::Error::CompiledTooBig(_)) => {
+            return Err(anyhow::Error::new(
+                crate::utils::limits::LimitError::RegexSizeLimitExceeded {
+                    limit: REGEX_SIZE_LIMIT,
+                },
+            ));
+        }
         // TODO: This behavior is due to OPA test not raising error. Should we raise error?
         _ => return Ok(Value::Undefined),
```
[the remark this answers is not in the document]
Comment (anakrish) [reply src/builtins/regex.rs:235]: Good catch — added `blocked-regex-replace-oversized` test case that exercises this exact bespoke `CompiledTooBig` branch with an oversized pattern, verifying it produces a "size limit" error.

File: src/rvm/vm/functions.rs [hunk 7]
```diff
@@ -122,7 +122,16 @@ impl RegoVM {
...
+            // Resource-limit errors must always propagate, even in non-strict
+            // mode, to prevent `not builtin(...)` from silently flipping to true.
+            Err(e) if !self.strict_builtin_errors => {
+                if e.downcast_ref::<crate::LimitError>().is_some() {
+                    self.dummy_exprs = dummy_exprs;
+                    self.cached_builtin_args = args;
+                    return Err(e.into());
+                }
+                Value::Undefined
+            }
```
[the remark this answers is not in the document]
Comment (anakrish) [reply src/rvm/vm/functions.rs:134]: Good point. However, threading `self.pc` into the error at this site would require changing the `From<anyhow::Error>` path to a manual conversion — the `.into()` call is idiomatic Rust error handling via the `?` operator. The pc here is the *builtin call instruction*, not the specific regex operation within it. Since the error message already identifies the exact limit violated and its value, the actionable fix for the user is always "simplify your regex pattern" regardless of pc. I will note this as a potential improvement but don't think it warrants a code change in this PR.

File: src/builtins/regex.rs [hunk 8]
```diff
@@ -147,11 +177,21 @@ fn find_n(span: &Span, params: &[Ref<Expr>], args: &[Value], _strict: bool) -> R
...
+    };
+    match get_or_compile_regex(&pattern) {
+        Ok(_) => Ok(Value::Bool(true)),
+        // Size-limit exceeded is a resource-limit violation; propagate as hard error.
+        Err(regex::Error::CompiledTooBig(_)) => Err(anyhow::Error::new(
+            crate::utils::limits::LimitError::RegexSizeLimitExceeded {
+                limit: REGEX_SIZE_LIMIT,
+            },
+        )),
+        // Syntax errors mean the pattern is genuinely invalid.
```
[the remark this answers is not in the document]
Comment (anakrish) [reply src/builtins/regex.rs:192]: This is an intentional design decision. `regex.is_valid` answers "can this pattern be used with the regex builtins?" — if the engine refuses to compile a pattern due to size limits, it is not valid *for use in this engine*. Returning `false` would be misleading: it implies the pattern has a syntax error, when in reality the engine is refusing to compile it for safety reasons.

Returning an error makes the failure visible and actionable. A policy that needs to probe arbitrary patterns can wrap the call in a rule body where the error makes the rule undefined — the standard Rego error-handling pattern.

This is consistent with how OPA's `regex.is_valid` behaves when their engine rejects a pattern for resource reasons.

File: src/builtins/regex.rs [hunk 9]
```diff
@@ -72,8 +104,7 @@ fn find_all_string_submatch_n(
     let value = ensure_string(name, &params[1], &args[1])?;
     let n = ensure_numeric(name, &params[2], &args[2])?;
 
-    let re = get_or_compile_regex(&pattern)
-        .or_else(|_| bail!(params[0].span().error("invalid regex")))?;
+    let re = compile_regex_for_builtin(params[0].span(), &pattern)?;
```
[the remark this answers is not in the document]
Comment (anakrish) [reply src/builtins/regex.rs:107]: Added test cases for `regex.find_all_string_submatch_n` and `regex.template_match` with oversized patterns in the next push.

File: src/builtins/regex.rs [hunk 10]
```diff
@@ -242,8 +287,10 @@ fn regex_template_match(
         }
 
         // Fetch pattern, excluding delimiters.
-        let re = get_or_compile_regex(&template[start + delimiter_start.len()..end])
-            .or_else(|_| bail!(params[0].span().error("invalid regex")))?;
+        let re = compile_regex_for_builtin(
+            params[0].span(),
+            &template[start + delimiter_start.len()..end],
+        )?;
```
[the remark this answers is not in the document]
Comment (anakrish) [reply src/builtins/regex.rs:293]: Will add in the same update — see response to the `find_all_string_submatch_n` comment.

File: src/interpreter.rs [hunk 11]
```diff
@@ -2405,8 +2405,14 @@ impl Interpreter {
...
-            // Ignore errors if we are not evaluating in strict mode.
-            Err(_) if !self.compiled_policy.strict_builtin_errors => return Ok(Value::Undefined),
+            // Resource-limit errors must always propagate, even in non-strict
+            // mode, to prevent `not builtin(...)` from silently flipping to true.
+            Err(e) if !self.compiled_policy.strict_builtin_errors => {
+                if e.downcast_ref::<crate::LimitError>().is_some() {
+                    return Err(e);
+                }
+                return Ok(Value::Undefined);
+            }
```
[the remark this answers is not in the document]
Comment (anakrish) [reply src/interpreter.rs:2415]: The interpreter path (`src/interpreter.rs:2415`) is exercised by the existing test infrastructure — `tests/interpreter/` runs the same YAML test cases through the tree-walking interpreter. The `regex_size_limit.yaml` file is picked up by both the RVM and interpreter test runners. The `not regex.match(...)` case specifically exercises this negation path.

I'll verify this is the case and confirm.

File: tests/rvm/rego/cases/regex_size_limit.yaml [hunk 12]
```diff
@@ -0,0 +1,285 @@
...
+  # regex.replace with oversized pattern must error
+  # -------------------------------------------------------------------------
+  - note: blocked-regex-replace-oversized
+    data: {}
+    modules:
+      - |
+        package test
+        main := regex.replace("hello", `([a-z]{1,64}){1,64}`, "x")
+    query: data.test.main
+    want_error: "size limit"
```
[the remark this answers is not in the document]
Comment (anakrish) [reply tests/rvm/rego/cases/regex_size_limit.yaml:285]: Good catch on the count — will update the PR description to reflect the actual 29 cases (13 blocked + 16 legitimate).

File: src/builtins/regex.rs [hunk 13]
```diff
@@ -147,11 +177,21 @@ fn find_n(span: &Span, params: &[Ref<Expr>], args: &[Value], _strict: bool) -> R
...
+        Err(_) => return Ok(Value::Bool(false)),
+    };
+    match get_or_compile_regex(&pattern) {
+        Ok(_) => Ok(Value::Bool(true)),
+        // Size-limit exceeded is a resource-limit violation; propagate as hard error.
+        Err(regex::Error::CompiledTooBig(_)) => Err(anyhow::Error::new(
+            crate::utils::limits::LimitError::RegexSizeLimitExceeded {
+                limit: REGEX_SIZE_LIMIT,
+            },
+        )),
```
[the remark this answers is not in the document]
Comment (anakrish) [reply src/builtins/regex.rs:191]: The duplication exists because `regex.replace` and `regex.is_valid` have different error-handling semantics from the other builtins:
- `regex.replace`: OPA returns `Undefined` for syntax errors but we want to hard-fail for size limits
- `regex.is_valid`: needs to distinguish syntax errors (→ `false`) from size limits (→ error)

The shared helper `compile_regex_for_builtin()` handles the common case. The two special cases require inline handling because their non-size-limit error behavior differs. The `LimitError::RegexSizeLimitExceeded` construction is a one-liner and the limit value comes from a single `const` — changing it requires updating one place.

File: src/builtins/regex.rs [hunk 14]
```diff
@@ -185,6 +224,13 @@ fn regex_replace(
 
     let re = match get_or_compile_regex(&pattern) {
         Ok(p) => p,
+        Err(regex::Error::CompiledTooBig(_)) => {
+            return Err(anyhow::Error::new(
+                crate::utils::limits::LimitError::RegexSizeLimitExceeded {
+                    limit: REGEX_SIZE_LIMIT,
+                },
+            ));
+        }
```
[the remark this answers is not in the document]
Comment (anakrish) [reply src/builtins/regex.rs:233]: See response on the sibling thread — the duplication is intentional due to differing error semantics in `regex.replace` vs. other builtins.

File: src/builtins/regex.rs [hunk 15]
```diff
@@ -21,6 +21,21 @@ use regex::Regex;
...
+/// comfortably, while adversarial patterns that would otherwise cause
+/// expensive DFA construction are rejected at compile time.
+const REGEX_SIZE_LIMIT: usize = 100 * 1024;
+
+/// Compile a regex pattern with a size limit to bound resource consumption.
+fn compile_regex(pattern: &str) -> core::result::Result<Regex, regex::Error> {
+    RegexBuilder::new(pattern)
+        .size_limit(REGEX_SIZE_LIMIT)
+        .build()
+}
```
[the remark this answers is not in the document]
Comment (anakrish) [reply src/builtins/regex.rs:37]: Good catch — the PR description was inaccurate. The limit is currently a hardcoded constant (`REGEX_SIZE_LIMIT = 100 * 1024`), which is intentional for this initial PR. Making it configurable at runtime (threading it through engine settings) can be added as a follow-up if deployments need to tune it. Updated the PR description to remove the misleading claim.

File: src/rvm/vm/functions.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (anakrish) [reply src/rvm/vm/functions.rs:134]: Same feedback as the earlier thread — `pc: 0` is a structural limitation of the `From<anyhow::Error>` conversion path which is triggered by the `?` operator. Threading `self.pc` would require replacing idiomatic `.into()` with a manual conversion at every builtin call site. The error message itself ("regex compiled size exceeds limit of 100KB") is the actionable diagnostic; the RVM instruction offset adds no user value for this class of error.

File: src/builtins/regex.rs [hunk 16]
```diff
@@ -41,10 +56,27 @@ fn get_or_compile_regex(pattern: &str) -> core::result::Result<Regex, regex::Err
...
+fn compile_regex_for_builtin(span: &Span, pattern: &str) -> Result<Regex> {
+    get_or_compile_regex(pattern).map_err(|e| match e {
+        regex::Error::CompiledTooBig(_) => {
+            anyhow::Error::new(crate::utils::limits::LimitError::RegexSizeLimitExceeded {
+                limit: REGEX_SIZE_LIMIT,
+            })
+        }
+        _ => anyhow::anyhow!(span.error("invalid regex")),
+    })
+}
```
[the remark this answers is not in the document]
Comment (anakrish) [reply src/builtins/regex.rs:78]: The duplication is a one-liner (`LimitError::RegexSizeLimitExceeded { limit: REGEX_SIZE_LIMIT }`) that exists because each site has different surrounding error-handling semantics:
- `compile_regex_for_builtin`: common path, returns `anyhow::Error`
- `regex.is_valid`: must distinguish syntax errors (→ `false`) from size limits (→ hard error)
- `regex.replace`: OPA-compatible `Undefined` for syntax errors but hard error for size limits

A shared mapper would need to know which error type to return for the non-size-limit case, making it more complex than the current approach. The limit value comes from a single `const` — changing it requires updating one place.

File: src/builtins/regex.rs [hunk 17]
```diff
@@ -147,11 +177,21 @@ fn find_n(span: &Span, params: &[Ref<Expr>], args: &[Value], _strict: bool) -> R
...
+        Ok(_) => Ok(Value::Bool(true)),
+        // Size-limit exceeded is a resource-limit violation; propagate as hard error.
+        Err(regex::Error::CompiledTooBig(_)) => Err(anyhow::Error::new(
+            crate::utils::limits::LimitError::RegexSizeLimitExceeded {
+                limit: REGEX_SIZE_LIMIT,
+            },
+        )),
+        // Syntax errors mean the pattern is genuinely invalid.
+        Err(_) => Ok(Value::Bool(false)),
+    }
```
[the remark this answers is not in the document]
Comment (anakrish) [reply src/builtins/regex.rs:194]: See response on the sibling thread at line 78.

File: src/builtins/regex.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (anakrish) [reply src/builtins/regex.rs:233]: See response on the sibling thread at line 78.

File: src/builtins/regex.rs [hunk 18]
```diff
@@ -118,8 +149,7 @@ fn find_n(span: &Span, params: &[Ref<Expr>], args: &[Value], _strict: bool) -> R
     let value = ensure_string(name, &params[1], &args[1])?;
     let n = ensure_numeric(name, &params[2], &args[2])?;
 
-    let re = get_or_compile_regex(&pattern)
-        .or_else(|_| bail!(params[0].span().error("invalid regex")))?;
+    let re = compile_regex_for_builtin(params[0].span(), &pattern)?;
```
[the remark this answers is not in the document]
Comment (anakrish) [reply src/builtins/regex.rs:152]: Added `blocked-find-n-oversized` test case covering `regex.find_n` with an oversized pattern.

File: src/builtins/regex.rs [hunk 19]
```diff
@@ -198,8 +244,7 @@ fn regex_split(span: &Span, params: &[Ref<Expr>], args: &[Value], _strict: bool)
     let pattern = ensure_string(name, &params[0], &args[0])?;
     let value = ensure_string(name, &params[1], &args[1])?;
 
-    let re = get_or_compile_regex(&pattern)
-        .or_else(|_| bail!(params[0].span().error("invalid regex")))?;
+    let re = compile_regex_for_builtin(params[0].span(), &pattern)?;
```
[the remark this answers is not in the document]
Comment (anakrish) [reply src/builtins/regex.rs:247]: Added `blocked-split-oversized` test case covering `regex.split` with an oversized pattern.

File: src/interpreter.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (anakrish) [reply src/interpreter.rs:2415]: The test harness at `tests/rvm/rego/mod.rs` runs every YAML case through **both** the tree-walking interpreter (`engine.eval_rule()` at line 366) and the RVM. For `want_error` cases, the harness asserts that both paths produce an error — if either succeeds, the test panics. So this interpreter fix IS exercised by the existing regression suite.

You can verify by checking the test output: cases like `blocked-not-regex-match-oversized` validate both engines.

File: src/rvm/vm/functions.rs [hunk 20]
```diff
@@ -122,7 +122,16 @@ impl RegoVM {
...
             Ok(value) => value,
-            Err(_) if !self.strict_builtin_errors => Value::Undefined,
+            // Resource-limit errors must always propagate, even in non-strict
+            // mode, to prevent `not builtin(...)` from silently flipping to true.
+            Err(e) if !self.strict_builtin_errors => {
+                if e.downcast_ref::<crate::LimitError>().is_some() {
+                    self.dummy_exprs = dummy_exprs;
+                    self.cached_builtin_args = args;
+                    return Err(e.into());
+                }
```
[the remark this answers is not in the document]
Comment (anakrish) [reply src/rvm/vm/functions.rs:132]: This is the same feedback raised in two prior review rounds (threads at `errors.rs:328` and `functions.rs:134`). Already addressed: `pc: 0` is a structural limitation of `From<anyhow::Error>` which is invoked by the `?` operator. The error message itself identifies the violated limit — the RVM instruction offset adds no user value for resource-limit errors. See prior thread for full rationale.

File: src/builtins/regex.rs [hunk 21]
```diff
@@ -41,10 +56,27 @@ fn get_or_compile_regex(pattern: &str) -> core::result::Result<Regex, regex::Err
...
+///   evaluator may swallow to `Undefined` in non-strict mode (OPA-compatible).
+fn compile_regex_for_builtin(span: &Span, pattern: &str) -> Result<Regex> {
+    get_or_compile_regex(pattern).map_err(|e| match e {
+        regex::Error::CompiledTooBig(_) => {
+            anyhow::Error::new(crate::utils::limits::LimitError::RegexSizeLimitExceeded {
+                limit: REGEX_SIZE_LIMIT,
+            })
+        }
+        _ => anyhow::anyhow!(span.error("invalid regex")),
+    })
```
[the remark this answers is not in the document]
Comment (anakrish) [reply src/builtins/regex.rs:77]: This is consistent with how all other `LimitError` variants (`TimeLimitExceeded`, `MemoryLimitExceeded`) propagate — none carry a source span. The error bubbles up through the interpreter call stack and the caller has the expression span available for diagnostic display. This is the standard error propagation pattern in the codebase; adding span only to `RegexSizeLimitExceeded` would be inconsistent with the other limit errors.

File: src/rvm/vm/rules.rs [hunk 22]
```diff
@@ -79,6 +109,13 @@ impl RegoVM {
                 {
                     match self.jump_to(destructuring_entry_point) {
                         Ok(_result) => {}
+                        Err(e) if Self::is_fatal_vm_error(&e) => {
+                            self.restore_rule_state(
+                                &mut previous_loop_stack,
+                                &mut previous_comprehension_stack,
+                            );
+                            return Err(e);
+                        }
```
[the remark this answers is not in the document]
Comment (anakrish) [reply src/rvm/vm/rules.rs:118]: This is a repeat of the comments at `rules.rs:119` and `rules.rs:158` from the first review round. The state cleanup was already fixed with `restore_rule_state()`. Testing the destructuring path specifically would require a contrived rule whose destructuring block invokes a regex builtin — this is not a realistic pattern in Rego policies. The end-to-end error propagation is tested through the existing `want_error` cases which exercise both interpreter and RVM paths.

File: src/utils/limits/error.rs [hunk 23]
```diff
@@ -61,6 +70,9 @@ impl fmt::Display for LimitError {
                     usage, limit
                 )
             }
+            Self::RegexSizeLimitExceeded { limit } => {
+                write!(f, "compiled regex exceeded size limit ({} bytes)", limit)
```
[the remark this answers is not in the document]
Comment (anakrish) [reply src/utils/limits/error.rs:74]: The message reads: "compiled regex exceeded size limit (102400 bytes)" — this clearly states that the regex exceeded the limit, and shows the configured cap in parentheses. This is the same pattern used by `TimeLimitExceeded` ("time limit exceeded ({limit}ms)") and `MemoryLimitExceeded` ("memory limit exceeded ({limit} bytes)"). The wording is consistent and unambiguous.

**Review by dpokluda — APPROVED** [review]
